### PR TITLE
pyspacer training: CN+MERMAID normalization & generalized rollups/filtering

### PR DESCRIPTION
**Pull request by StephenChan**
(Description update) This PR got pretty big, but here's what's in it:

## Ability to use CoralNet training data, not just MERMAID

- CoralNet sources to include are specified with a CSV input file, one source ID per row. (Per previous discussions, ability to specify an exclusions list instead can be nice to have later, but not top priority for now.)
    
- The CN to MERMAID label mapping is drawn from the MERMAID API. I imitated Viktor's code to retrieve the mapping page by page.
    
- The annotation limit option now tries to balance out the annotations between CoralNet and MERMAID. Also, this option no longer has extra complexity which forces taking either all points or no points (as opposed to a subset) from each image, since this is just an option for testing/sanity-checks anyway.

## Rollups and filtering
    
- Generalize rollups by making it not just BAs to roll up to, but BA+GF combos to roll up from and roll up to.

- Generalize filtering by making it not just BAs to include/exclude, but BA+GF combos to include/exclude.

- The `CsvSpec` class in `common/csv_utils.py` achieves a measure of DRY between the various CSV inputs to training.

- The ability to accept either IDs or names from rollup, inclusion, and exclusion CSVs has been left out for now; since it was looking a bit complicated with all the extra columns being added, and I just wanted to wrap up this PR. The CSVs can still include BA/GF names, but they just won't be parsed in any way for now, and IDs are mandatory. If name-only CSVs are still deemed useful (like for quick tests), we can add that capability back soon.

## MLflow artifacts

- Add an artifact listing CoralNet labels that had to be excluded from the data because they don't map to anything in MERMAID. The artifact includes the number of sources and number of annotations that the label was found in (among the CSV-specified CN sources).
    
- Add an artifact for the API's full CoralNet-MERMAID label mapping at the time of training.
    
- Add artifacts for the number of images and annotations from each 'project'. One artifact lists the 'raw' numbers before rollups and filtering. Another artifact lists only the annotations left after rollups and filtering, including how much went into the train/ref/val sets. Since we aren't getting information about which individual MERMAID projects the data are from, all MERMAID data is grouped under a single project called 'all'. CoralNet sources are each individual projects.
    
- In the BA counts and BAGF counts, add the number of projects contributing to each count, and the number of annotations going into train/ref/val. There is also a column for number of 'dropped' annotations: the train/ref/val splitting logic (the core of which is done in sklearn, not our code) seems to drop annotations for rare BAGFs whose annotations would be spread too thin after stratification. (I think it happens if any of the three sets would end up with 0 annotations, but I'm not 100% sure.)
    
- Optionally log a tabular artifact with one row per individual annotation that was eligible for training, including each annotation's train/ref/val/dropped category. Options are no annotations (default), all annotations (can be a big artifact), annotations from a specific CoralNet source, or annotations from a specific CoralNet image. More options could be added later if useful. Overall, this could be useful as a sanity check about what's going into training.
    
- The artifact variables in the code are now a bit less scattered, as `TrainingDataset.artifacts` now provides a namespace for most of them.
    
## DuckDB

I figured that either DuckDB or pandas could take us down the route of making the for-loops leaner, and I ended up primarily using DuckDB since (to my understanding) that is the option which could also lead to lighter RAM usage (though not necessarily at this moment; I haven't measured).

DuckDB is now used to implement rollups, filtering, combining CN+MERMAID data, counting annotations, building feature vector paths, and more.

This all drew some patterns and inspiration from Lauren's IC v1 code. Additionally, there were a number of DuckDB patterns I tried to come up with myself, though I suspect some of those patterns could be done with significantly fewer steps by anyone who's more familiar with DuckDB (or even SQL generally). The only other times I've written this much raw SQL were when working with PHP/MySQL websites from the early 2000s, and I'm not sure how I feel about that. At any rate, I made some of these patterns reusable (and thus, potentially wholesale improvable) in `common/duckdb_utils.py`.

## Functionality related to S3 access

- The `train-data-source-flexibility` pyspacer feature branch is now used so that we can draw train data from multiple S3 buckets. This change is needed so that we can temporarily read CoralNet data from a separate bucket, as CN isn't set up to use the open data bucket yet. This change will eventually be in the main branch of pyspacer, but for now we can use this feature branch, which AFAIK doesn't have outstanding issues for this repo's purposes.
    
- AWS auth can now be done by specifying a key, secret, and token in a .env file. The motivation for this is that the aforementioned temporary bucket with CoralNet data isn't public, and this auth method provides a way of accessing said data from a local machine outside of AWS (which is useful for my coding, at least). For the record, my understanding is that AWS-MERMAID SSO gives keys that expire after 12 hours.
    
- The AWS region, as well as S3 bucket names for CoralNet and MERMAID, are now configurable in .env. Again, largely motivated by the CoralNet data bucket not being public right now, but we've been wanting to have these things configurable anyway.

### CODE REVIEW

**Comment by ircaldwell** [comment]
Looks amazing. Thank you for all this Stephen.

**Review by saanobhaai — APPROVED** [review]
